### PR TITLE
Add logging for pipeline

### PR DESCRIPTION
**Pull request by Moishe**
This PR adds some visual logging to watch frames go through pipelines.

Here's an example from sample 04. This sample uses two Pipelines with a SequentialMergePipeline, so the hardcoded one runs first. I *think* this makes it easy to see which frames are being passed along.

Note that I drop Audio frames. I might change this to show the first and then an ellipse or something until there's something that's *not* an audio frame.

```
10 2024-03-27 18:46:04,228   source -> TextFrame: "My friend the LLM is going to tell a joke about llamas" -> AzureTTSService
10 2024-03-27 18:46:04,229     AzureTTSService -> TextFrame: "My friend the LLM is going to tell a joke about llamas" -> AzureTTSService
10 2024-03-27 18:46:04,229   source -> EndPipeFrame -> AzureTTSService
10 2024-03-27 18:46:04,229     AzureTTSService -> EndPipeFrame -> AzureTTSService
10 2024-03-27 18:46:04,229       AzureTTSService -> TTSStartFrame -> sink
20 2024-03-27 18:46:04,229 🎬 Starting frame consumer thread
20 2024-03-27 18:46:04,229 Running azure tts
10 2024-03-27 18:46:04,230   source -> LLMMessagesQueueFrame -> ElevenLabsTTSService
10 2024-03-27 18:46:04,230     ElevenLabsTTSService -> LLMMessagesQueueFrame -> AzureLLMService
10 2024-03-27 18:46:04,230       AzureLLMService -> LLMResponseStartFrame -> ElevenLabsTTSService
10 2024-03-27 18:46:04,230         ElevenLabsTTSService -> LLMResponseStartFrame -> sink
10 2024-03-27 18:46:04,230 Generating chat via openai: [{"content": "tell the user a joke about llamas", "role": "system", "name": "system"}]
20 2024-03-27 18:46:04,863 === OpenAI LLM TTFB: 0.6326563358306885
20 2024-03-27 18:46:05,412 Got azure tts result
20 2024-03-27 18:46:05,412 Returning result
10 2024-03-27 18:46:05,412       AzureTTSService -> TTSEndFrame -> sink
10 2024-03-27 18:46:05,412       AzureTTSService -> TextFrame: "My friend the LLM is going to tell a joke about llamas" -> sink
10 2024-03-27 18:46:05,412       AzureTTSService -> EndPipeFrame -> sink
10 2024-03-27 18:46:05,827       AzureLLMService -> TextFrame: "Why" -> ElevenLabsTTSService
10 2024-03-27 18:46:05,828       AzureLLMService -> TextFrame: " don" -> ElevenLabsTTSService
10 2024-03-27 18:46:05,829       AzureLLMService -> TextFrame: "'t" -> ElevenLabsTTSService
10 2024-03-27 18:46:05,829       AzureLLMService -> TextFrame: " ll" -> ElevenLabsTTSService
10 2024-03-27 18:46:05,829       AzureLLMService -> TextFrame: "amas" -> ElevenLabsTTSService
10 2024-03-27 18:46:05,829       AzureLLMService -> TextFrame: " ever" -> ElevenLabsTTSService
10 2024-03-27 18:46:05,830       AzureLLMService -> TextFrame: " get" -> ElevenLabsTTSService
10 2024-03-27 18:46:05,830       AzureLLMService -> TextFrame: " lost" -> ElevenLabsTTSService
10 2024-03-27 18:46:05,830       AzureLLMService -> TextFrame: "?

" -> ElevenLabsTTSService
10 2024-03-27 18:46:05,830         ElevenLabsTTSService -> TTSStartFrame -> sink
10 2024-03-27 18:46:06,515         ElevenLabsTTSService -> TTSEndFrame -> sink
10 2024-03-27 18:46:06,515         ElevenLabsTTSService -> TextFrame: "Why don't llamas ever get lost?

" -> sink
10 2024-03-27 18:46:06,516       AzureLLMService -> TextFrame: "Because" -> ElevenLabsTTSService
10 2024-03-27 18:46:06,516       AzureLLMService -> TextFrame: " they" -> ElevenLabsTTSService
10 2024-03-27 18:46:06,516       AzureLLMService -> TextFrame: " always" -> ElevenLabsTTSService
10 2024-03-27 18:46:06,517       AzureLLMService -> TextFrame: " take" -> ElevenLabsTTSService
10 2024-03-27 18:46:06,517       AzureLLMService -> TextFrame: " the" -> ElevenLabsTTSService
10 2024-03-27 18:46:06,518       AzureLLMService -> TextFrame: " llama" -> ElevenLabsTTSService
10 2024-03-27 18:46:06,518       AzureLLMService -> TextFrame: "-n" -> ElevenLabsTTSService
10 2024-03-27 18:46:06,518       AzureLLMService -> TextFrame: "ated" -> ElevenLabsTTSService
10 2024-03-27 18:46:06,519       AzureLLMService -> TextFrame: " route" -> ElevenLabsTTSService
10 2024-03-27 18:46:06,519       AzureLLMService -> TextFrame: "!" -> ElevenLabsTTSService
10 2024-03-27 18:46:06,519         ElevenLabsTTSService -> TTSStartFrame -> sink
10 2024-03-27 18:46:07,194         ElevenLabsTTSService -> TTSEndFrame -> sink
10 2024-03-27 18:46:07,194         ElevenLabsTTSService -> TextFrame: "Because they always take the llama-nated route!" -> sink
10 2024-03-27 18:46:07,197       AzureLLMService -> LLMResponseEndFrame -> ElevenLabsTTSService
10 2024-03-27 18:46:07,198         ElevenLabsTTSService -> LLMResponseEndFrame -> sink
10 2024-03-27 18:46:07,198   source -> EndPipeFrame -> ElevenLabsTTSService
10 2024-03-27 18:46:07,198     ElevenLabsTTSService -> EndPipeFrame -> AzureLLMService
10 2024-03-27 18:46:07,198       AzureLLMService -> EndPipeFrame -> ElevenLabsTTSService
10 2024-03-27 18:46:07,198         ElevenLabsTTSService -> EndPipeFrame -> sink
20 2024-03-27 18:46:13,816 Stopping frame consumer thread
```

### CODE REVIEW

File: src/dailyai/pipeline/pipeline.py [hunk 1]
```diff
@@ -96,18 +101,32 @@ async def run_pipeline(self):
...
                 ):
                     yield final_frame
         else:
             yield initial_frame
+
+    def _log_frame(self, frame: Frame, depth: int):
+        if not isinstance(frame, AudioFrame):
+            if depth == 0:
+                source = "source"
+                dest = str(self._processors[depth - 1])
```
Comment (aconchillo) [inline src/dailyai/pipeline/pipeline.py:124]: if `depth` is 0 will the index be -1?

File: src/dailyai/pipeline/pipeline.py [hunk 2]
```diff
@@ -96,18 +101,32 @@ async def run_pipeline(self):
...
+            elif depth == len(self._processors) + 1:
+                dest = "sink"
+                source = str(self._processors[depth - 2])
+            else:
+                dest = str(self._processors[depth - 1])
+                source = str(self._processors[depth - 2])
```
Comment (aconchillo) [inline src/dailyai/pipeline/pipeline.py:130]: same here. if `depth` is 1 will the index be -1?

Comment (Moishe) [reply src/dailyai/pipeline/pipeline.py]: Argh, you're right. Fixed and added tests.

**Comment by chadbailey59** [comment]
To be clear, is this logging frames as they are "between" services?

```
AzureLLMService -> TextFrame: "Why" -> ElevenLabsTTSService
```

This log line means "the AzureLLMService just emitted a TextFrame with the contents 'Why', and that frame is about to go into the ElevenLabsTTSService", right?

**Comment by chadbailey59** [comment]
I really like the approach here, but one other thought for readability's sake: In addition to the "depth" approach, Could we assign a color to each service? Maybe we only emit the escape codes if we're connected to a tty or something. But I'm envisioning the frames themselves stay as white text, but we build an array of colors that correspond to the 'depth', which I think would make it even easier to scan the logs when you're debugging.

**Review by chadbailey59 — APPROVED** [review]

**Comment by Moishe** [comment]
You're correct that each line is Frame Processor Yields -> Frame -> Frame Processor consumes.

The color thing is a cool idea -- I think I'll merge this and then maybe play with that in a separate PR. If we used a consistent color for each Frame Processor, maybe that could be more readable. I'll need to make sure that we can detect that the logger destination supports ANSI escape codes, too.

**Pull request merged**